doc/graphene-specificatin-v1.mediawiki

```diff
@@ -0,0 +1,323 @@
+|vData||<nowiki>vector<unsigned char></nowiki>||variable||Bit array for filter
+|-
+|isFull||bool||1 byte||True if every bit in vData is set to 1
+|-
+|isEmpty||bool||1 byte||True if every bit in vData is set to 0
+|-
+|nHashFuncs||unsigned int||2, 4 bytes||Number of hash functions used in filter
+|-
+|nTweak||unsigned int||2, 4, bytes||Additive offset for hash function inputs
+|-
+|nFlags||unsigned char||1 byte||Defines behavior for transaction insertion
+|}
+
+The standard Bloom filter implementation does not always honor the false positive rate (FPR) requested during initialization. This behavior is problematic for Graphene because the FPR determines the number of erroneous transactions removed from IBLT ''I'', which affects the decode rate for ''I''. For this reason, it is required that ''nDesiredSize'' and ''nHashFuncs'' be defined slightly differently than in the standard ''CBloomFilter'' implementation, as follows.
+
+<code>
+nDesiredSize = (unsigned int)(ceil(-1 / LN2SQUARED * nElements * log(nFPRate) / 8));
+</code>
+<br />
+<code>
+nHashFuncs = (unsigned int)max(MIN_N_HASH_FUNC, int(vData.size() * 8 / nElements * LN2));
+</code>
+
+====CIblt====
+''CIblt'' is an IBLT implementation that is part of the ''CGrapheneSet'' reconciliation process. In the absence of a reconciliation failure, ''CIblt'' will generate false positive list ''F'' and missing transaction cheap hashes ''M''.
+
+{|
+|'''Field Name'''||'''Type'''||'''Size'''||'''Encoding Details'''||'''Purpose'''
+|-
+|version||uint64_t||8 bytes||Compact size||Version bits
+|-
+|n_hash||uint8_t||1 byte||Standard||Number of hash functions used
+|-
+|is_modified||bool||1 byte||Standard||True if any items have been inserted into the IBLT
+|-
+|hashTable||<nowiki>vector<HashTableEntry></nowiki>||variable||Standard||Data cells for IBLT
+|}
+
+The Graphene protocol requires that IBLT '' I' '', created by the receiver, be subtracted from IBLT ''I'', originating from the sender. In order for this operation to succeed, it is critical that the IBLTs use the same quantity of hash functions, have the same number of cells, and that hash function ''i'' uses the same function and seed for both ''I'' and '' I' ''. Although the sender may use [https://github.com/umass-forensics/IBLT-optimization complex optimization techniques] to determine the number of cells and hash functions for ''I'', the receiver should simply copy those values provided that they are reasonably sized. The ''i''th hash function is the [http://code.google.com/p/smhasher/source/browse/trunk/MurmurHash3.cpp MurmurHash3] with seed ''i''.
+
+====HashTableEntry====
+''HashTableEntry'' implements a single cell of the ''CIblt'' data structure.
+
+{|
+|'''Field Name'''||'''Type'''||'''Size'''||'''Purpose'''
+|-
+|count||uint32_t||4 bytes||Number of items
+|-
+|keySum||uint64_t||8 bytes||XOR of all keys
+|-
+|keyCheck||uint32_t||4 bytes||Error checking for keySum
+|-
+|valueSum||vector<nowiki><uint8_t></nowiki>||variable||XOR of all values
```

Comment:
> valueSum||vector<nowiki><uint8_t></nowiki>||variable||XOR of all values

This field is unused and adds one byte of overhead to each table entry for no purpose.  For a set difference of size 600 and factor of 1.4 as used in examples here, this adds 840 bytes of zero bytes.  (For comparison, in Bitcoin I'm able to get >99% block reproduction using fewer bytes than this one source of overhead alone, so it's not insubstantial...)

Reply:
Thanks for pointing this out. We did actually have this optimization in mind in the [original paper](https://people.cs.umass.edu/~gbiss/graphene.pdf) and @awemany also pointed it out. But we decided to leave the IBLT as general-purpose for the first version and are planning optimizations like this for following versions. The spec here was drawn up to match the existing implementation, i.e. it does not detail future improvements. Of course suggestions are welcome nevertheless.

Reply:
For this kind of usage it shouldn't even have a count field at all, it's pure loss at the targeted failure rate; but instead should choose a 'checksum' that will be collision free for two colliding elements.   From other research: https://people.xiph.org/~greg/temp/hybrid-miniclub.png  The graph is all in overhead for coding 30-bit elements, at a really high failure rate (1:100) the differences are much larger at lower failure rates (I just graphed at 1:100 because it takes less cpu time...)

LOL-iblt in that graph is essentially your current code (only distinction is 30-bit elements), legacy is one from the cpisync repo (which shares a common heritage with your code but is somewhat different), opt iblt is a basic iblt without any of the obvious mistakes that common implementations, including yours, seem to have...  Hybrid is some ongoing research that combines iblt and minisketch, though at the 1:100 failure rate level it's not earth shattering.  All the IBLT approaches are _really_ inefficient compared to minisketch, however.

In any case, I think it's really disappointing that your spec doesn't mention the triviality of colliding it and address handling that case.  Even BIP152, which doesn't have that vulnerability, still addresses it more directly.